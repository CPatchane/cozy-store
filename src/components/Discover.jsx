import React, { Component } from 'react'

import { translate } from 'cozy-ui/react/I18n'

import SmallAppItem from './SmallAppItem'

const DEMO_APPS = [
  {
    slug: 'bank',
    name: 'Bank',
    developer: {name: 'Cozy'},
    editor: 'Cozy',
    icon: require('../assets/temp_apps/icon-bank.svg')
  },
  {
    slug: 'sante',
    name: 'Sant√©',
    developer: {name: 'MAIF Ltd'},
    editor: '',
    icon: require('../assets/temp_apps/icon-sante.svg')
  },
  {
    slug: 'mon_logis',
    name: 'Mon Logis',
    developer: {name: 'HoodBrains'},
    editor: '',
    icon: require('../assets/temp_apps/icon-mon_logis.svg')
  },
  {
    slug: 'calendar',
    name: 'Calendar',
    developer: {name: 'Cozy'},
    editor: 'Cozy',
    icon: require('../assets/temp_apps/icon-calendar.svg')
  }
]

class Discover extends Component {
  render () {
    const { t } = this.props
    return (
      <div className='sto-discover sto-soon'>
        <h2 className='sto-discover-title'>{t('discover.title')}</h2>
        <div className='sto-discover-get-started'>
          <h3 className='sto-discover-get-started-title'>
            {t('discover.get_started')}
          </h3>
          <div className='sto-discover-get-started-list'>
            {DEMO_APPS.map(a => {
              return <SmallAppItem
                slug={a.slug}
                developer={a.developer}
                editor={a.editor}
                icon={a.icon}
                name={a.name}
                version={a.version}
              />
            })}
          </div>
        </div>
      </div>
    )
  }
}

export default translate()(Discover)
