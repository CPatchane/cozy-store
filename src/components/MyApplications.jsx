import React, { Component } from 'react'

import { Route } from 'react-router-dom'
import { translate } from 'cozy-ui/react/I18n'
import Spinner from 'cozy-ui/react/Spinner'

import SmallAppItem from './SmallAppItem'
import ApplicationModal from './ApplicationModal'

class MyApplications extends Component {
  constructor (props) {
    super(props)
    props.fetchApps()
  }

  onAppClick (appSlug) {
    this.props.history.push(`/myapps/${appSlug}/manage`)
  }

  render () {
    const { t, myApps, isFetching, error } = this.props
    return (
      <div className='sto-myapps'>
        <h2 className='sto-myapps-title'>{t('myapps.title')}</h2>
        <div className='sto-myapps-list'>
          {!isFetching && myApps && !!myApps.length &&
            myApps.map(a => {
              return <SmallAppItem
                slug={a.slug}
                developer={a.developer}
                editor={a.editor}
                icon={a.icon}
                name={a.name}
                version={a.version}
                onClick={() => this.onAppClick(a.slug)}
              />
            })
          }
          {error &&
            <p>{error}</p>
          }
        </div>
        {isFetching &&
          <Spinner
            size='xxlarge'
            loadingType='appsFetching'
            middle='true'
          />
        }

        <Route path='/myapps/:appSlug/manage' render={({ match }) =>
          (<ApplicationModal {...this.props} match={match} />)
        } />
      </div>
    )
  }
}

export default translate()(MyApplications)
