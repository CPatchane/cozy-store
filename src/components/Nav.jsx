import React from 'react'
import { translate } from 'cozy-ui/react/I18n'
import { NavLink } from 'react-router-dom'

export const Nav = ({ t }) => (
  <nav>
    <ul className='coz-nav'>
      <li className='coz-nav-item'>
        <NavLink to='/discover' className='sto-cat-discover coz-nav-link' activeClassName='active'>
          { t('nav.discover') }
        </NavLink>
      </li>
      <li className='coz-nav-item'>
        <NavLink to='/myapps' className='sto-cat-myapps coz-nav-link' activeClassName='active'>
          { t('nav.myapps') }
        </NavLink>
      </li>
    </ul>
  </nav>
)

export default translate()(Nav)
