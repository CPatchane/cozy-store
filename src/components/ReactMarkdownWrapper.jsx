import React from 'react'
import ReactMarkdown from 'react-markdown'

export const reactMarkdownRendererOptions = {
  Link: props => <a href={props.href} target='_blank'>{props.children}</a>
}

export const ReactMarkdownWrapper = ({ source }) =>
  <ReactMarkdown
    source={source}
    renderers={reactMarkdownRendererOptions}
  />

export default ReactMarkdownWrapper
