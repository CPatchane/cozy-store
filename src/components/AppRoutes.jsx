import React from 'react'
import { Route, Switch, Redirect } from 'react-router-dom'

import MyApplications from '../containers/MyApplications'

const AppRoutes = () => {
  return <Switch>
    <Route path='/myapps' component={MyApplications} />
    <Redirect from='/' to='/myapps' />
  </Switch>
}

export default AppRoutes
