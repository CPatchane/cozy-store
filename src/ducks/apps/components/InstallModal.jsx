import React, { Component } from 'react'

import { translate } from 'cozy-ui/react/I18n'
import Modal, { ModalContent } from 'cozy-ui/react/Modal'

import PermissionsList from './PermissionsList'
import ReactMarkdownWrapper from '../../components/ReactMarkdownWrapper'

export class InstallModal extends Component {
  constructor (props) {
    super(props)
    this.state = {
      error: null
    }

    if (props.app) props.fetchLastAppVersion(props.app.slug)

    this.gotoParent = this.gotoParent.bind(this)
    this.installApp = this.installApp.bind(this)
    this.getPermissions = this.getPermissions.bind(this)
  }

  installApp () {
    this.setState({ error: null })
    const { app } = this.props
    this.props.installApp(app.slug)
    .then(() => {
      this.gotoParent()
    })
    .catch(error => {
      this.setState({ error })
    })
  }

  gotoParent () {
    const { parent } = this.props
    this.context.router.history.push(parent)
  }

  getPermissions () {
    const { currentAppVersion, t } = this.props
    const manifest = currentAppVersion && currentAppVersion.manifest
    const permissionsArray = Object.values(manifest.permissions) || []
    return permissionsArray.map(p => {
      p.typeDescription = (t(`doctypes.${p.type}`)).replace(/^doctypes\./, '')
      return p
    })
  }

  render () {
    const { t, app, isVersionFetching, currentAppVersion, versionError, isInstalling } = this.props
    const { error } = this.state
    // if app not found, return to parent
    if (!app) {
      this.gotoParent()
      return null
    }
    let permissions = null
    if (currentAppVersion && !isVersionFetching && !versionError) permissions = this.getPermissions()
    return (
      <div className='sto-modal--install'>
        <Modal
          secondaryAction={this.gotoParent}
        >
          <ModalContent>
            <header className='sto-modal-header'>
              <div className='sto-modal-header-icon' aria-busy={isVersionFetching}>
                <a href='https://cozy.io' target='_blank' title='Cozy Website' class='sto-modal-header-icon-shield' />
              </div>
              <h2>{t('app_modal.install.title', {appName: app.name})}</h2>
            </header>
            <div className='sto-modal-content'>
              {permissions && <PermissionsList permissions={permissions} appName={app.name} />
              }
              {!isVersionFetching && !versionError &&
                <div>
                  {permissions && !!permissions.length &&
                    <ReactMarkdownWrapper
                      source={t('app_modal.install.accept_description', {
                        appName: app.name
                      })
                      }
                    />
                  }
                  <div className='sto-modal-controls'>
                    <button
                      role='button'
                      className='coz-btn coz-btn--secondary'
                      onClick={this.gotoParent}
                    >
                      {t('app_modal.install.cancel')}
                    </button>
                    <button
                      role='button'
                      disabled={isInstalling}
                      aria-busy={isInstalling}
                      className='coz-btn coz-btn--regular coz-btn--download'
                      onClick={this.installApp}
                    >
                      {t('app_modal.install.install')}
                    </button>
                  </div>
                </div>
              }
              {versionError &&
                <p class='coz-error'>{t('app_modal.install.message.version_error', {message: versionError.message})}</p>
              }
              {error &&
                <p class='coz-error'>{t('app_modal.install.message.install_error', {message: error.message})}</p>
              }
            </div>
          </ModalContent>
        </Modal>
      </div>
    )
  }
}

export default translate()(InstallModal)
