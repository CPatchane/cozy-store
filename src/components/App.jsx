
import React from 'react'

import Sidebar from './Sidebar'
import AppRoutes from './AppRoutes'

const App = () => (
  <div className='sto-wrapper coz-sticky'>
    <Sidebar />
    <main className='sto-content'>
      <AppRoutes />
    </main>
  </div>
)
export default App
