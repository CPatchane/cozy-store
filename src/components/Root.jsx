import React from 'react'
import { Provider } from 'react-redux'
import { I18n } from 'cozy-ui/react/I18n'
import { HashRouter } from 'react-router-dom'

import App from './App'

const Root = ({ context, lang, store }) => {
  return <I18n lang={lang} dictRequire={(lang) => require(`../locales/${lang}`)}>
    <Provider store={store}>
      <HashRouter>
        <App />
      </HashRouter>
    </Provider>
  </I18n>
}

export default Root
