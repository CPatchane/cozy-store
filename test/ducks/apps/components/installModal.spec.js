'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { InstallModal } from 'ducks/apps/components/InstallModal'

import mockApps from '../_mockApps'
import mockAppVersion from '../_mockPhotosRegistryVersion'

/* SinonJS is used here to stub Promise in order to be synchronous.
In this way, (p)React will call setState synchronously. It will allow
to assert the component state juste after */
import sinon from 'sinon'
import sinonStubPromise from 'sinon-stub-promise'
sinonStubPromise(sinon)

const mockError = new Error('This is a test error')

const getMockProps = (fromRegistry = false) => ({
  app: fromRegistry
    ? Object.assign(
        {},
        mockAppVersion.manifest,
        mockApps.find(a => a.slug === 'photos')
      )
    : mockApps.find(a => a.slug === 'photos'),
  dismissAction: jest.fn(),
  parent: '/discover',
  onAlreadyInstalled: jest.fn(),
  onSuccess: jest.fn(app => {
    if (app.slug === 'photos' || app.slug === 'konnector-bouilligue') {
      return sinon
        .stub()
        .returnsPromise()
        .resolves(mockApps.find(a => a.slug === app.slug))()
    }
    return sinon
      .stub()
      .returnsPromise()
      .rejects(mockError)()
  }),
  history: {
    push: jest.fn()
  }
})

describe('InstallModal component', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  beforeAll(() => {
    // define global mock url
    global.cozy = {
      client: {
        _url: '//cozytest.mock.cc'
      }
    }
  })

  it('should be rendered correctly if app found', () => {
    const component = shallow(<InstallModal {...getMockProps()} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly with also app in registry', () => {
    const mockProps = getMockProps(true)
    const component = shallow(<InstallModal {...mockProps} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly to update app with availableVersion', () => {
    const mockProps = getMockProps(true)
    mockProps.app.installed = true
    mockProps.app.availableVersion = '4.0.0'
    const component = shallow(<InstallModal {...mockProps} />).getElement()
    expect(component).toMatchSnapshot()
    expect(mockProps.onAlreadyInstalled.mock.calls.length).toBe(0)
  })

  it('should not break the permissions part if no permissions property found in manifest', () => {
    const mockProps = getMockProps(true)
    delete mockProps.app.permissions
    const component = shallow(<InstallModal {...mockProps} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should handle focus trop correctly', () => {
    const wrapper = shallow(<InstallModal {...getMockProps()} />)
    expect(wrapper.state().activeTrap).toBe(true)
    wrapper.instance().unmountTrap()
    expect(wrapper.state().activeTrap).toBe(false)
  })

  it('calls onAlreadyInstalled if app installed without update', () => {
    const mockProps = getMockProps(true)
    mockProps.app.installed = true
    shallow(<InstallModal {...mockProps} />)
    expect(mockProps.onAlreadyInstalled.mock.calls.length).toBe(1)
  })

  it('calls onAlreadyInstalled if app status updated to installed', () => {
    const mockProps = getMockProps(true)
    mockProps.app.installed = false
    const wrapper = shallow(<InstallModal {...mockProps} />)
    expect(mockProps.onAlreadyInstalled.mock.calls.length).toBe(0)
    mockProps.app.installed = true
    wrapper.setProps(mockProps)
    expect(mockProps.onAlreadyInstalled.mock.calls.length).toBe(1)
  })
})
