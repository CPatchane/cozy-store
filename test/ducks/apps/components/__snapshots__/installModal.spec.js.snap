// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstallModal component should be rendered correctly if app found 1`] = `
<div
  className="sto-modal--install"
>
  <Modal
    overflowHidden={false}
    primaryType="secondary"
    secondaryAction={[Function]}
    secondaryType="regular"
    withCross={true}
  >
    <ModalContent>
      <header
        className="sto-modal-header"
      >
        <div
          aria-busy={undefined}
          className="sto-modal-header-icon"
        >
          <a
            class="sto-modal-header-icon-shield"
            href="https://cozy.io"
            target="_blank"
            title="Cozy Website"
          />
        </div>
        <h2>
          The application Photos will not transmit any personal data to any third parties.
        </h2>
      </header>
      <div
        className="sto-modal-content"
      >
        <_translate
          appName="Photos"
          permissions={Object {}}
        />
        <div>
          <div
            className="sto-modal-controls"
          >
            <button
              className="coz-btn coz-btn--secondary"
              onClick={[Function]}
              role="button"
            >
              Cancel
            </button>
            <button
              aria-busy={undefined}
              className="coz-btn coz-btn--regular coz-btn--download"
              disabled={undefined}
              onClick={[Function]}
              role="button"
            >
              Install
            </button>
          </div>
        </div>
      </div>
    </ModalContent>
  </Modal>
</div>
`;

exports[`InstallModal component should be rendered correctly with also app in registry 1`] = `
<div
  className="sto-modal--install"
>
  <Modal
    overflowHidden={false}
    primaryType="secondary"
    secondaryAction={[Function]}
    secondaryType="regular"
    withCross={true}
  >
    <ModalContent>
      <header
        className="sto-modal-header"
      >
        <div
          aria-busy={undefined}
          className="sto-modal-header-icon"
        >
          <a
            class="sto-modal-header-icon-shield"
            href="https://cozy.io"
            target="_blank"
            title="Cozy Website"
          />
        </div>
        <h2>
          The application Photos will not transmit any personal data to any third parties.
        </h2>
      </header>
      <div
        className="sto-modal-content"
      >
        <_translate
          appName="Photos"
          permissions={
            Object {
              "albums": Object {
                "description": "Required to manage photos albums",
                "methods": Array [
                  "GET",
                  "POST",
                  "PUT",
                ],
                "type": "io.cozy.photos.albums",
              },
              "apps": Object {
                "description": "Required by the cozy-bar to display the icons of the apps",
                "type": "io.cozy.apps",
                "verbs": Array [
                  "GET",
                ],
              },
              "contacts": Object {
                "description": "Required to to share photos with your contacts",
                "methods": Array [
                  "GET",
                  "POST",
                ],
                "type": "io.cozy.contacts",
              },
              "files": Object {
                "description": "Required for photo access",
                "methods": Array [
                  "GET",
                  "POST",
                  "PUT",
                ],
                "type": "io.cozy.files",
              },
              "settings": Object {
                "description": "Required by the cozy-bar to display Claudy and know which applications are coming soon",
                "type": "io.cozy.settings",
                "verbs": Array [
                  "GET",
                ],
              },
            }
          }
        />
        <div>
          <Unknown
            source="By electing __Install__, you accept that the application Photos obtains access to these data, only locally and within your Cozy, and in compliance with the terms of service and the privacy policy."
          />
          <div
            className="sto-modal-controls"
          >
            <button
              className="coz-btn coz-btn--secondary"
              onClick={[Function]}
              role="button"
            >
              Cancel
            </button>
            <button
              aria-busy={undefined}
              className="coz-btn coz-btn--regular coz-btn--download"
              disabled={undefined}
              onClick={[Function]}
              role="button"
            >
              Install
            </button>
          </div>
        </div>
      </div>
    </ModalContent>
  </Modal>
</div>
`;

exports[`InstallModal component should handle correctly fetchError from props 1`] = `
<div
  className="sto-modal--install"
>
  <Modal
    overflowHidden={false}
    primaryType="secondary"
    secondaryAction={[Function]}
    secondaryType="regular"
    withCross={true}
  >
    <ModalContent>
      <header
        className="sto-modal-header"
      >
        <div
          aria-busy={undefined}
          className="sto-modal-header-icon"
        >
          <a
            class="sto-modal-header-icon-shield"
            href="https://cozy.io"
            target="_blank"
            title="Cozy Website"
          />
        </div>
      </header>
      <div
        className="sto-modal-content"
      >
        <p
          class="coz-error"
        >
          Something went wrong when fetching the application informations. Reason: This is a test error
        </p>
      </div>
    </ModalContent>
  </Modal>
</div>
`;

exports[`InstallModal component should handle correctly installError from props 1`] = `
<div
  className="sto-modal--install"
>
  <Modal
    overflowHidden={false}
    primaryType="secondary"
    secondaryAction={[Function]}
    secondaryType="regular"
    withCross={true}
  >
    <ModalContent>
      <header
        className="sto-modal-header"
      >
        <div
          aria-busy={undefined}
          className="sto-modal-header-icon"
        >
          <a
            class="sto-modal-header-icon-shield"
            href="https://cozy.io"
            target="_blank"
            title="Cozy Website"
          />
        </div>
        <h2>
          The application Photos will not transmit any personal data to any third parties.
        </h2>
      </header>
      <div
        className="sto-modal-content"
      >
        <_translate
          appName="Photos"
          permissions={Object {}}
        />
        <div>
          <p
            class="coz-error"
          >
            Something went wrong when installing the application on your Cozy. Reason: This is a test error
          </p>
          <div
            className="sto-modal-controls"
          >
            <button
              className="coz-btn coz-btn--secondary"
              onClick={[Function]}
              role="button"
            >
              Cancel
            </button>
            <button
              aria-busy={undefined}
              className="coz-btn coz-btn--regular coz-btn--download"
              disabled={undefined}
              onClick={[Function]}
              role="button"
            >
              Install
            </button>
          </div>
        </div>
      </div>
    </ModalContent>
  </Modal>
</div>
`;

exports[`InstallModal component should not break the permissions part if no permissions property found in manifest 1`] = `
<div
  className="sto-modal--install"
>
  <Modal
    overflowHidden={false}
    primaryType="secondary"
    secondaryAction={[Function]}
    secondaryType="regular"
    withCross={true}
  >
    <ModalContent>
      <header
        className="sto-modal-header"
      >
        <div
          aria-busy={undefined}
          className="sto-modal-header-icon"
        >
          <a
            class="sto-modal-header-icon-shield"
            href="https://cozy.io"
            target="_blank"
            title="Cozy Website"
          />
        </div>
        <h2>
          The application Photos will not transmit any personal data to any third parties.
        </h2>
      </header>
      <div
        className="sto-modal-content"
      >
        <_translate
          appName="Photos"
          permissions={Object {}}
        />
        <div>
          <div
            className="sto-modal-controls"
          >
            <button
              className="coz-btn coz-btn--secondary"
              onClick={[Function]}
              role="button"
            >
              Cancel
            </button>
            <button
              aria-busy={undefined}
              className="coz-btn coz-btn--regular coz-btn--download"
              disabled={undefined}
              onClick={[Function]}
              role="button"
            >
              Install
            </button>
          </div>
        </div>
      </div>
    </ModalContent>
  </Modal>
</div>
`;
